### ISSUE TITLE
Unsupported keyword: timedate.

### ISSUE THREAD

**Issue by aharjula** [maths/moodle-qtype_stack#284]
(Some data analysis related tasks have been in demand lately and this comes from them, we either need to generate data or be able to take data in in bulk form.)

Trying to generate a random dataset with actual dates without having to deal with the number of days in a given month and though that might as well use `timedate()` for it but did not get far. Any reason why it would be forbidden? Also if I wanted to extract a numerical number of month (or any other bit) from it how should I do that as I can't use `parse_string` and I know very well that I will never get to use it?

Chris, do you know about some way of parsing integers or numbers in general from "strings" inside Maxima? And do you think we should maybe provide such a thing (though not using parse_string) or maybe even directly a function that would return lists of dates in numeric form for this kind of an use case?

```
startDate: 3385045277; /* seconds since midnight 19000101*/
endDate: startDate + 3*365*24*60*60;

dataSetB: [];
dumvar: for iterDate:startDate step 24*60*60 while iterDate < endDate do (r:split(split(timedate(iterDate)," ")[1],"-"),r:append(r,[rand(50)+30,rand(20)+20]),dataSetB:append(dataSetB,[r]));
```

The only reason I could come up for forbidding `timedate()` is that someone might start using it as a source for seeds for randomisation and not understand that that would cause major problems. Is there any other reasons for it to be forbidden?

**Comment by sangwinc** [comment]
Matti,

There is no special reason why timedate() is forbidden.  A long time ago I went through everything which looks like it calls the operating system and forbade it.  That is all.

I suggest we create a modest suite of time & date functions which you need.  There is no problem in using parse_string in the main STACK libraries in a tied down way.  This sounds like a sensible use, so if you have specific proposals we should accommodate them.

Chris

**Comment by sangwinc** [comment]
Matti,

What about adding something like this to the core STACK code?
`split_timedate():= block([S1, S],
    S1:timedate(),
    /* Gives us something like this: "2017-05-02 17:20:06+01:00". */
    S:split(S1, "+"),
    S:flatten([split(first(S), " "), second(S)]),
    S:flatten([split(first(S), "-"), split(second(S), ":"), split(third(S), ":")]),
    S:map(parse_string, S),
    S:append([S1, parse_timedate(S1)], S)
)$`

**Comment by sangwinc** [comment]
@aharjula What about adding something like the following to the main STACK code and letting teachers use this?

<code>
split_timedate():= block([S1, S2, S],
    S1:timedate(),
    /* Gives us something like this: "2017-05-02 17:20:06+01:00", or "2017-05-02 17:20:06-01:00". */
    S1:"2017-05-02 17:20:06-03:00",
    S:split(S1, " "),
    S:flatten([split(first(S), "-"), map(lambda([ex], split(ex, ":") ), rest(S))]),
    S2:S[6],
    if cequal(charat(S[6], 3), "+") then 
        S[6]:[substring(S2,1,3), "\"+\"", substring(S2,4)]
    else 
        S[6]:[substring(S2,1,3), "\"-\"", substring(S2,4)],
    S:flatten(S),
    S:map(parse_string, S),
    S:append([S1, parse_timedate(S1)], S)
)$
</code>

**Comment by aharjula** [comment]
Not far from what is needed, although that example itself is something we do not want i.e. something depending on the current time, imagine the caching of that and time-variant questions... Naturally, in the real case you give it the timedate to split.

My wish list after some thought is:
 1. Basically, that function, although should probably just return year, month and day for a given timedate. And if at all possible weekday as a number, but maxima does not seem to have it, day of year might also be handy. This would be the tool that tells you what is the date 2^15 seconds from a given date.
 2. A convenience date sequence generation function that you give a start date (inclusive) and an end date (not inclusive) and it returns a list of the first functions return values for that range. This would probably be a six integer parameter function, but strings are also good. This function would be the one that would give a list of dates in a given years February if you can't be bothered to count the leap year yourself or just a sequence for generating a dataset tied to dates.
 3. There might be a long version of the first one that would return time as well as date, but I don't really see the point for that unless maxima handles DST and leap seconds perfectly and based on the fact that weekday was not easily found I would guess this too will be problematic.

**Comment by sangwinc** [comment]
Matti,
Perhaps you would like to propose some specific Maxima functions and we will put these in a near future version?

**Comment by aharjula** [comment]
This has now been fixed in v4.0 with the functions:

- time_to_date(seconds) => [year, month, day, weekday]
- day_for_date(year, month, day) => weekday
- date_list(yearA, monthA, dayA, yearB, monthB, dayB) => [[yearA, monthA, dayA, weekdayA],[]...]  list of lists describing the dates between the given dates not including the second date.
